Release powershell v2

pdoelwijt opened the pull request.

Rick-Jongbloed committed:
Remove version nr

Rick-Jongbloed committed:
First version of V2 Create

Rick-Jongbloed committed:
New version of configuration definition

Rick-Jongbloed committed:
final version before testing

Rick-Jongbloed committed:
Local updates

pdoelwijt committed:
Initial release

Initial release

rschouten97 commented on create.ps1:
```diff
@@ -1,42 +1,32 @@
 
-# Set TLS to accept TLS, TLS 1.1 and TLS 1.2
-[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls -bor [Net.SecurityProtocolType]::Tls11 -bor [Net.SecurityProtocolType]::Tls12
+#$MutationDate = $actionContext.Data.beginDatum.Substring(0,10) # startdate from contract
```
I assume these are settings used for testing. If that’s the case, please clean this up. If not, adjust the code to correctly calculate mutationdate.

pdoelwijt replied:
fixed

rschouten97 commented on create.ps1:
```diff
@@ -101,278 +89,108 @@ function Get-ErrorMessage {
+        Write-Verbose 'Correlating user account'
+        $uri = "$($actionContext.Configuration.BaseUri)/medewerkers/$($currentAccount.Id)/$($MutationDate)"
+        $bodyobject = [PSCustomObject]@{}
+        $bodyobject | Add-Member -MemberType NoteProperty -Name EmailZakelijk -Value $actionContext.Data.EmailZakelijk -Force
```
I recommend defining an object, such as $account, and mapping it with the data. This approach makes the connector more dynamic, eliminating the need for script changes when adding or modifying field mappings. While not critical, it enhances the connector’s flexibility.

rschouten97 commented on create.ps1:
```diff
@@ -101,278 +89,108 @@ function Get-ErrorMessage {
+    # Process
+    if (-not($actionContext.DryRun -eq $true)) {
+        
+        Write-Verbose 'Correlating user account'
```
Please modify this process to correlate only with the user account. Currently, the update action occurs during correlation. However, HelloID automatically handles the update when the account is correlated and the variable $outputContext.AccountCorrelated is set to true. To align with this behavior, remove the update from the create action and move it to the update script.

pdoelwijt replied:
fixed

rschouten97 commented on create.ps1:
```diff
@@ -101,278 +89,108 @@ function Get-ErrorMessage {
+        $outputContext.success = $true
+        $outputContext.AccountCorrelated = $true
+        $outputContext.AuditLogs.Add([PSCustomObject]@{
+                Action  = 'CreateAccount'
```
The action should be “CorrelateAccount” since you only correlate to the account, and no create action is performed.

pdoelwijt replied:
fixed

rschouten97 commented on delete.ps1:
```diff
@@ -24,7 +24,7 @@ $WarningPreference = "Continue"
+    CorrelationValue    = $aRef # Has to match the value of the unique identifier
 }
 
 # Change mapping here
```
Please modify the mapping to no longer define $updateAccountFields and $storeAccountFields in the script. Instead, these should be defined in the field mapping using the Ps v2 connector.

rschouten97 commented on update.ps1:
```diff
@@ -1,43 +1,34 @@
 
-# Set TLS to accept TLS, TLS 1.1 and TLS 1.2
-[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls -bor [Net.SecurityProtocolType]::Tls11 -bor [Net.SecurityProtocolType]::Tls12
+#$MutationDate = $actionContext.Data.beginDatum.Substring(0,10) # startdate from contract
```
I assume these are settings used for testing. If that’s the case, please clean this up. If not, adjust the code to correctly calculate mutationdate.

pdoelwijt replied:
fixed

rschouten97 commented on update.ps1:
```diff
@@ -232,29 +196,26 @@ try {
+                    $outputContext.success = $true
+                    $outputContext.AuditLogs.Add([PSCustomObject]@{
                             Message = "Successfully updated account [$($currentAccount.RoepNaam) $($currentAccount.AchterNaam) ($($updatedAccount.Id))]. Updated properties: $($changedPropertiesObject | ConvertTo-Json -Depth 10)"
+                            #Message = 'Update account was successful'
```
I assume this is also used for testing. If that’s the case, please clean this up. If not, adjust the code to use the correct message.

pdoelwijt replied:
removed second message

rschouten97 reviewed (changes requested):
Overall, the code looks good. However, a few changes are needed before we can publish this. If the connector has been successfully tested and implemented at a client, I recommend thoroughly testing the changes there to ensure the connector functions as expected.

rschouten97 commented on configuration.json:
```diff
@@ -45,13 +45,13 @@
   },
   {
-    "key": "updateOnCorrelate",
+    "key": "updateOnUpdate",
```
This key has been changed or added, but nowhere in the scripts has it been used. You have two options: either modify the scripts to check and utilize this key, or, if it’s not needed, remove it from the configuration.

pdoelwijt replied:
fixed, removed the keys

pdoelwijt committed:
updates due to pull request review

rschouten97 committed:
removed version 1 from headers

rschouten97 committed:
Removed variable $storeAccountFields as this is moved to fieldmapping

rschouten97 committed:
Removed variable $MutationDate as this is not used

rschouten97 committed:
Cleaned up formatting

rschouten97 committed:
Removed variable $auditLogs as this is available by default as $outputContext.AuditLogs

rschouten97 committed:
Removed variable $storeAccountFields as this is moved to fieldmapping

rschouten97 committed:
Cleaned up formatting

rschouten97 committed:
Defined variable $account, and mapped it with data

rschouten97 committed:
Added $MutationDate as it is still required

rschouten97 committed:
Removed dump of currentaccount

rschouten97 committed:
Cleaned up formatting

rschouten97 committed:
Updated to match template

rschouten97 committed:
Updated to match template

rschouten97 committed:
Updated to match template

rschouten97 committed:
Updated to match template

rschouten97 committed:
Updated to match updated scripts

rschouten97 committed:
Update update.ps1

rschouten97 committed:
Update update.ps1

rschouten97 committed:
Update delete.ps1

rscholtelubberink committed:
Small fix in account reference

Small fix in account reference, as it is set to references.account and  not references.account.id

rhouthuijzen commented on create.ps1:
```diff
@@ -1,378 +1,222 @@
 
 # Set debug logging
-switch ($($c.isDebug)) {
+switch ($actionContext.Configuration.isDebug) {
```
If making chances maybe remove isDebug and replace all write-verbose for write-information?
Also this can be removed then:
$VerbosePreference = "SilentlyContinue"
$InformationPreference = "Continue"
$WarningPreference = "Continue"

rhouthuijzen commented on delete.ps1:
```diff
@@ -1,343 +1,321 @@
-    #region Get current account and verify if there are changes
-    try {
-        Write-Verbose "Querying account where [$($correlation.CorrelationProperty)] = [$($correlation.CorrelationValue)]"
+    Write-Verbose "Created headers. Result: $($headers | ConvertTo-Json)."
```
We shouldn't print the header.

rhouthuijzen commented on delete.ps1:
```diff
@@ -1,343 +1,321 @@
 
-        $splatWebRequest = @{
-            Uri             = "$($c.BaseUri)/medewerkersbasic/$($correlation.CorrelationValue)"
+    if ($actionContext.CorrelationConfiguration.Enabled) {
```
Why is this option in the delete script?

rhouthuijzen commented on delete.ps1:
```diff
@@ -1,343 +1,321 @@
+        # SDBHR docs: https://api.sdbstart.nl/swagger/ui/index#!/Medewerkers/Medewerkers_GetMedewerkerV2
+
+        $getSDBHRAccountSplatParams = @{
+            Uri             = "$($actionContext.Configuration.BaseUri)/medewerkersbasic/$correlationValue"
```
Shouldn't $correlationValue be accountRefference?

rhouthuijzen commented on update.ps1:
```diff
@@ -1,334 +1,312 @@
-    #region Get current account and verify if there are changes
-    try {
-        Write-Verbose "Querying account where [$($correlation.CorrelationProperty)] = [$($correlation.CorrelationValue)]"
+    Write-Verbose "Created headers. Result: $($headers | ConvertTo-Json)."
```
We shouldn't print the header.

rhouthuijzen commented on update.ps1:
```diff
@@ -1,334 +1,312 @@
+    Write-Verbose "Created headers. Result: $($headers | ConvertTo-Json)."
+    #endregion Create headers
+
+    if ($actionContext.CorrelationConfiguration.Enabled) {
```
Why is this in the update script?

rhouthuijzen commented on update.ps1:
```diff
@@ -1,334 +1,312 @@
-        $splatWebRequest = @{
-            Uri             = "$($c.BaseUri)/medewerkersbasic/$($correlation.CorrelationValue)"
+        $getSDBHRAccountSplatParams = @{
+            Uri             = "$($actionContext.Configuration.BaseUri)/medewerkersbasic/$correlationValue"
```
Shouldn't $correlationValue be accountRefference?

rhouthuijzen commented on delete.ps1:
```diff
@@ -1,343 +1,321 @@
+            $actionMessage = "updating account"
+        
+            # Throw terminal error
+            throw "No account found where [$($correlationField)] = [$($correlationValue)] while this connector only supports correlation."
```
This should be action skipped (no error)

rhouthuijzen commented on delete.ps1:
```diff
@@ -1,343 +1,321 @@
+            $actionMessage = "updating account"
+    
+            # Throw terminal error
+            throw "Multiple accounts found where [$($correlationField)] = [$($correlationValue)]. Please correct this so the persons are unique."
```
This should be action skipped (no error)

rhouthuijzen commented on configuration.json:
```diff
@@ -44,16 +44,6 @@
-    }
-  },
   {
     "key": "isDebug",
```
Maybe remove?

rhouthuijzen reviewed (commented):
Overall code looks fine. I made some recommendations and I'm not sure if this is ever tested.

rhouthuijzen committed:
Fix: fixes after review
